### ISSUE TITLE
Disable reverse DNS resolution (PTR lookup)

### ISSUE THREAD

**Issue by pevik** [iputils/iputils#531]
`ping` in iputils has always been doing reverse DNS resolution (PTR lookup) by default.

It's possible to disable it with `-n` (which can be set as alias), but people have been asking to disable it (https://github.com/iputils/iputils/pull/485#pullrequestreview-1590508258, https://github.com/iputils/iputils/issues/421#issuecomment-1327953699, https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=650478, https://unix.stackexchange.com/questions/400351/what-are-the-differences-between-iputils-ping-and-inetutils-ping/549830?noredirect=1#comment1474716_549830).

We hesitated to change due backwards compatibility, but because we have now [`-H`](https://github.com/iputils/iputils/pull/494), we *could* change. To be more friendly to the old behavior, we could introduce some variable, which would restore the old behavior without having to use `-H`). And obviously mention the change and the variable in man page and in release notes.

And we even have reverse DNS resolution bug, which I was not able to reproduce: https://github.com/iputils/iputils/issues/414.

@iputils/maintainers WDYT?

Other `ping` implementations (checked both code and `tcpdump` output):

* [`fping`](https://fping.org/) has it off by default + have `-d`, `--rdns` show targets by name (force reverse-DNS lookup), `-n`, `--name` show targets by name (reverse-DNS lookup for target IPs)
* [`ping` from busybox](https://git.busybox.net/busybox/tree/networking/ping) does not do reverse lookup and don't have option for it (calls `getnameinfo()` always with `NI_NUMERICHOST`).
* [`ping` from inetutils](https://git.savannah.gnu.org/cgit/inetutils.git/tree/ping) does it only for IPv6, but for each ICMP6 ECHO REQUEST. (Similar to https://github.com/iputils/iputils/issues/414 but only for IPv4, maybe bug?)

**Comment by nmeyerhans** [comment]
Other modern replacements for older tools (e.g. `ss`, as a replacement for `netstat`) disable DNS resolution by default.  I think it's reasonable to consider doing this in ping as well. I like the idea of using an environment variable to re-enable the old behavior, as it will simplify the transition for the many places where ping is scripted.

**Comment by pevik** [comment]
Thanks a lot for your input.

I wonder what should be the variable name? `IPUTILS_PING_DO_REVERSE_DNS` or `IPUTILS_PING_DO_PTR_LOOKUP` ?

We do reverse lookup also on `tracepath`, should should we follow the same approach? If yes, the variable  might be generic (without `PING`). Or are we OK with `-n` as most of the others (see below)?

Comparison (for aussiebroadband.com.au):
* `tracepath` (iputils): 12 searches, 2 searches with `-n`
* [`mtr`](https://github.com/traviscross/mtr): 10 searches, 2 searches with `-n`
* [`traceroute` (Dmitry Butskoy)](http://traceroute.sourceforge.net/): only 2 searches (no `-n` option)
* [`traceroute` from Busybox](https://git.busybox.net/busybox/tree/networking/traceroute.c): 13 searches, 2 searches with `-n`
* [`traceroute` from Pietrasanta (based on Dmitry Butskoy)](https://github.com/catchpoint/Networking.traceroute/) ([see Fosdem 2024](https://fosdem.org/2024/schedule/event/fosdem-2024-2929--where-the-are-the-packets-going-/)): 13 searches, 2 searches with `-n`.

**Comment by zhangyoufu** [comment]
> We do reverse lookup also on tracepath, should should we follow the same approach?

I think reverse lookup output is generally more useful than pure IP address in `tracepath`. And there are some traceroute poetry that may be affected if we change the default behavior of `tracepath`. For example, `tracepath -m 60 bad.horse`.

IMHO, the real UX problem is **human users expects quick response**, which is impossible because both `ping` and `tracepath` are **designed to output in a linear manner** and **reverse lookup takes time**. We can turn off reverse lookup by default in `ping` to satisfy user expectation, but we can't do the same in `tracepath` because reverse lookup result is useful.

**Comment by yvs2014** [comment]
just to note: iputils-ping reverse DNS lookup depends on TARGET form too.
If it's in numerical form, DNS lookup is already disabled by default (at least for IPv4): `ping 1.1` for example

**Comment by rpigott** [comment]
Hi.

No strong opinion on what the default behavior should be, but FWIW I reckon that a lot of the impetus for removing rdns by default is actually caused by some sd-resolved bugs that unfortunately have made rdns lookups very slow with the default configuration. See [1][2] and consider the significant impact they have on the ux of ping for many distros that use this resolver with the upstream defaults.

When affected by either issue, ping can easily take ~10s to print the first line. Both are now fixed upstream and expected in the next release of systemd. With these fixes, I expect that many users will quickly go back to not knowing or caring about rdns in ping.

[1] systemd/systemd#28166
[2] systemd/systemd#31826

**Comment by pevik** [comment]
One of the issues mentions our issue: https://github.com/iputils/iputils/issues/491. While these problems are 1 year old, first report from https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=650478 with slowness due rDNS was in 2011, people asked also in https://github.com/iputils/iputils/issues/421#issuecomment-1327953699 in 2022. Therefore I suspect problems will be sooner or later back.

**Comment by yvs2014** [comment]
> How about simply not blocking on the PTR lookup?

Maybe because it boils down to sufficient code complications: with blocking `getnameinfo(3)/gethostbyaddr(3)` it needs multithreading, otherwise it requires to construct a correspondent DNS query and send it to a NS via non-blocking socket and poll a response

**Comment by pevik** [comment]
Exactly, have non blocking resolving would require to handle `getnameinfo()` in own thread, which is quite major change (more or less complete rewrite). 

I hopefully one day find enough time to rewrite ping - (codebase is very old asking for complete rewrite, maybe even in rust, maybe just implement part of the functionality (number of switches is awesome - nearly whole alphabet is taken and it's a question if all functionality is still relevant). But that would require to cover current ping with real tests to be sure the new implementation works more or less the same.

**Comment by yvs2014** [comment]
Rewritting unix tools like a ping (in rust and/or with multithreading) will kill its simplicity, and so that the use as a base utility. All the more that, one's need of extra functionality can be done with user wrappers. (Imho of course)

**Comment by pevik** [comment]
I'm getting off topic here discussing rust in this topic. Let's move it to separate ticket: https://github.com/iputils/iputils/issues/537.

**Comment by pevik** [comment]
In the end I prepared PR, which adds environment variable, but PRT lookup is disabled by default. I hope it's good enough to be merged before upcoming new release.

https://github.com/iputils/iputils/pull/553
